### PR TITLE
feat(installer,platform): support installing apps from charts

### PR DESCRIPTION
**Pull request by MaDongdong99**
feat(installer,platform): support installing apps from charts

/kind feature
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: cmd/tke-installer/app/installer/application.go [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+import (
+	"context"
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	applicationv1 "tkestack.io/tke/api/application/v1"
+	"tkestack.io/tke/cmd/tke-installer/app/config"
+	"tkestack.io/tke/cmd/tke-installer/app/installer/constants"
+)
+
+const (
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/application.go:30]: please move const to const.go

Comment (leoryu) [reply cmd/tke-installer/app/installer/application.go:30]: and remove duplicate

File: cmd/tke-installer/app/installer/application.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+	return nil
+}
+
+func (t *TKE) installApplications(ctx context.Context) error {
+
+	if len(t.Config.PlatformApps) == 0 {
+		return nil
+	}
+
+	// TODO: (workaround ) client init will only be called in createCluster.
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/application.go:95]: Is there any case which skip create global cluster?

File: cmd/tke-installer/app/installer/application.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+func (t *TKE) installApplication(ctx context.Context, platformApp config.PlatformApp, installedApps []applicationv1.App) error {
+
+	chart := platformApp.Chart
+
+	var found bool
+	var duplicated bool
+	for _, installedApp := range installedApps {
+		// if there's an existed application with the same namespace+name, we consider it as already exists
+		if chart.Name == installedApp.Spec.Name && chart.TargetNamespace == installedApp.Namespace {
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/application.go:129]: chart.Name is diff with app.spec.name, should be platformApp.name == installed.App.Spec.name.

File: cmd/tke-installer/app/installer/application.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+		if chart.Name == installedApp.Spec.Chart.ChartName && chart.Version == installedApp.Spec.Chart.ChartVersion {
+			duplicated = true
+			break
+		}
+	}
+	if found {
+		t.log.Infof("application already exists. we don't override applications while installing. %v/%v", chart.TargetNamespace, chart.Name)
+		return nil
+	}
+	if duplicated {
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/application.go:143]: I don't think a chart deploy in 2 ns as 2 apps should return err.

File: cmd/tke-installer/app/installer/application.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+func (t *TKE) installApplication(ctx context.Context, platformApp config.PlatformApp, installedApps []applicationv1.App) error {
+
+	chart := platformApp.Chart
+
+	var found bool
+	var duplicated bool
+	for _, installedApp := range installedApps {
+		// if there's an existed application with the same namespace+name, we consider it as already exists
+		if chart.Name == installedApp.Spec.Name && chart.TargetNamespace == installedApp.Namespace {
+			found = true
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/application.go:130]: found value is not necessary, should add log info and return here.

File: cmd/tke-installer/app/installer/application.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	"context"
+	"fmt"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	applicationv1 "tkestack.io/tke/api/application/v1"
+	"tkestack.io/tke/cmd/tke-installer/app/config"
+	"tkestack.io/tke/cmd/tke-installer/app/installer/constants"
+)
+
+const (
+	applicationInstallNamespaceDefault      = "default"
```
Comment (leoryu) [inline cmd/tke-installer/app/installer/application.go:31]: please use "k8s.io/apimachinery/pkg/apis/meta/v1"  NamespaceDefault instead

File: cmd/tke-installer/app/installer/application.go (hunk 1 above)
Comment (MaDongdong99) [reply cmd/tke-installer/app/installer/application.go]: moved

File: cmd/tke-installer/app/installer/application.go (hunk 3 above)
Comment (MaDongdong99) [reply cmd/tke-installer/app/installer/application.go]: fixed

File: cmd/tke-installer/app/installer/application.go (hunk 4 above)
Comment (MaDongdong99) [reply cmd/tke-installer/app/installer/application.go]: drop this branch

File: cmd/tke-installer/app/installer/application.go (hunk 5 above)
Comment (MaDongdong99) [reply cmd/tke-installer/app/installer/application.go]: removed

File: cmd/tke-installer/app/installer/application.go (hunk 6 above)
Comment (MaDongdong99) [reply cmd/tke-installer/app/installer/application.go]: instead

**Comment by leoryu** [comment]
LGTM

**Pull request merged**